src/imp/kill.c

```diff
@@ -0,0 +1,236 @@
+};
+
+/*  Hard-coded path to systemd cgroup mount directory. This may
+ *   need to be moved to configuration at some point.
+ */
+static const char cgroup_mount_dir[] = "/sys/fs/cgroup/systemd";
+
+/*  Return the systemd cgroup path for PID `pid` in the provided buffer
+ *  Looks up the 'name=systemd'[*] subsystem relative cgroup path in
+ *   /proc/PID/cgroups and prepends `cgroup_mount_dir` to get the
+ *   full path.
+ *
+ *  [*] perhaps could also use the "pids" cgroup.
+ */
+static int pid_systemd_cgroup_path (pid_t pid, char *buf, int len)
+{
+    int rc = -1;
+    FILE *fp;
+    size_t size = 0;
+    int n;
+    char file [4096];
+    char *line = NULL;
+
+    n = snprintf (file, sizeof(file), "/proc/%ju/cgroup", (uintmax_t) pid);
+    if ((n < 0) || (n >= (int) sizeof(file))
+        || !(fp = fopen (file, "r")))
+        return -1;
+
+    while ((n = getline (&line, &size, fp)) >= 0) {
+        char *nl;
+        char *relpath = NULL;
+        char *subsys = strchr (line, ':');
+        if ((nl = strchr (line, '\n')))
+            *nl = '\0';
+        if (subsys == NULL || *(++subsys) == '\0'
+            || !(relpath = strchr (subsys, ':')))
+            continue;
+        /* Nullify subsys, relpath is already nul-terminated at newline */
+        *(relpath++) = '\0';
+        if (strcmp (subsys, "name=systemd") == 0) {
+            n = snprintf (buf, len, "%s%s", cgroup_mount_dir, relpath);
+            if ((n > 0) && (n < len))
+                rc = 0;
+            break;
+        }
+    }
+
+    free (line);
+    fclose (fp);
+    return rc;
+}
+
+/*  return the file owner of 'path'
+ */
+static uid_t path_owner (const char *path)
+{
+    struct stat st;
+    if (stat (path, &st) < 0)
+        return ((uid_t) -1);
+    return st.st_uid;
+}
+
+static struct pid_info *pid_info_create (pid_t pid)
+{
+    struct pid_info *pi = calloc (1, sizeof (*pi));
+    if (pi == NULL)
+        return NULL;
+    if (pid < 0)
+        pid = -pid;
+    pi->pid = pid;
+    if (pid_systemd_cgroup_path (pid, pi->cg_path, sizeof (pi->cg_path)) < 0)
+        goto err;
+    if ((pi->cg_owner = path_owner (pi->cg_path)) == (uid_t) -1)
+        goto err;
+    return pi;
+err:
+    free (pi);
+    return NULL;
+}
+
+/*  Return true if the user executing the IMP is allowed to run
+ *   'flux-imp kill'. This is the same set of users allowed to run
+ *   'flux-imp exec', so look in exec.allowed-users.
+ */
+static bool imp_kill_allowed (const cf_t *conf)
+{
+    struct passwd * pwd = getpwuid (getuid ());
+    const cf_t *exec = cf_get_in (conf, "exec");
+
+    if (pwd && exec)
+        return cf_array_contains (cf_get_in (exec, "allowed-users"),
+                                  pwd->pw_name);
+    return false;
+}
+
+static void check_and_kill_process (struct imp_state *imp, pid_t pid, int sig)
+{
+    uid_t user = getuid ();
+    struct pid_info *p = NULL;
+
+    if (!imp_kill_allowed (imp->conf))
+        imp_die (1, "kill command not allowed");
```

Comment:
This might be an uninformed question:  does this code need an explicit check to avoid signaling root processes like the exec subcommand has?

Reply:
Hm, it shouldn't come up unless the instance is able to start a root process. If the cgroup in which the process is running is not owned by current uid, then `imp kill` will refuse to kill it.

Additionally, the actual userid of processes isn't checked at this point.

Reply:
Sounds good to me, thanks,